clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/customtemplates/TemplatesManager.kt

```diff
@@ -42,4 +47,46 @@ internal class TemplatesManager(templates: Collection<CustomTemplate>) {
     }
 
     fun isTemplateRegistered(templateName: String): Boolean = customTemplates.contains(templateName)
+
+    fun presentTemplate(notification: CTInAppNotification, inAppListener: InAppListener) {
+        val context = createContextFromInApp(notification, inAppListener) ?: return
+        val template = customTemplates[context.templateName]
+        if (template == null) {
+            logger.info("[CustomTemplates] Cannot find template with name ${context.templateName}")
+            return
+        }
+
+        when (val presenter = template.presenter) {
```

Comment:
can we capture presenter type and context in generics rather than using frequent `is` checks?

Reply:
Presenter and Context are generic for CustomTemplate, but we store and treat CustomTemplates regardless of their inner type (function or template, and the actual types of presenter and context) in all other places. The type checks are needed only when using the presenters, which is not pretty, but should only happen in the TemplateManager.